### PR TITLE
Create taxon query type

### PR DESCRIPTION
**Pull request by igorkrz**
In this PR new query type is created - query by taxon which will be visible on every collection just like Sylius latest product or Sylius taxon product

### CODE REVIEW

**Comment by emodric** [comment]
I'm not sure I understand the point of the query? "Query by taxon" - Query for what, products?

**Comment by igorkrz** [comment]
@emodric query for taxon and its children - this should be used in collections just like other query types already implemented

File: bundle/Resources/config/query_types.yaml [hunk 1]
```diff
@@ -3,3 +3,5 @@ query_types:
         name: 'Sylius latest products'
     sylius_taxon_products:
         name: 'Sylius taxon products'
+    sylius_taxon:
```
Comment (RandyCupic) [inline bundle/Resources/config/query_types.yaml]: @igorkrz Maybe this should be called differently (eg. `Sylius taxons`) as it returns multiple taxons? cc @emodric

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+     */
+    public function __construct(
+        private readonly TaxonRepositoryInterface $taxonRepository,
+        private readonly RequestStack $requestStack,
+    ) {}
+
+    public function buildParameters(ParameterBuilderInterface $builder): void
+    {
+        $builder->add(
+            'use_current_taxon',
```
Comment (RandyCupic) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: @igorkrz @emodric Maybe also this should be called differently, so that it's more clear what is this being used for?

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            return [];
+        }
+
+        $taxon = $this->getTaxon($query);
+
+        if (!$taxon instanceof TaxonInterface) {
+            return [];
+        }
+
+        $taxonCode = $taxon->getCode();
```
Comment (RandyCupic) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: @igorkrz I think that code is mandatory on Taxons?

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            $taxonSlug = trim($currentRequest->attributes->get('slug') ?? '');
+            if ($taxonSlug === '') {
+                return null;
+            }
+
+            return $this->taxonRepository->findOneBySlug($taxonSlug, $currentRequest->getLocale());
+        }
+
+        $taxonId = $query->getParameter('taxon_id')->getValue();
+        if (trim($taxonId ?? '') === '') {
```
Comment (RandyCupic) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: @igorkrz Is this needed? I think that there's a way to check if parameter is empty. cc @emodric

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        );
+    }
+
+    /**
+     * @return iterable<array-key, TaxonInterface>
+     */
+    public function getValues(Query $query, int $offset = 0, ?int $limit = null): iterable
+    {
+        $currentRequest = $this->requestStack->getCurrentRequest();
+        if (!$currentRequest instanceof Request) {
```
Comment (RandyCupic) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: @igorkrz Why is this necessary?

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        /** @var TaxonInterface[] $children */
+        $children = $this->taxonRepository->findChildren($taxonCode);
+
+        return $children;
+    }
+
+    public function getCount(Query $query): int
+    {
+        $currentRequest = $this->requestStack->getCurrentRequest();
+        if (!$currentRequest instanceof Request) {
```
Comment (RandyCupic) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: @igorkrz Why is this necessary?

File: lib/Item/ValueLoader/TaxonValueLoader.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    public function load($id): ?TaxonInterface
+    {
+        try {
+            $taxon = $this->taxonRepository->find($id);
+        } catch (Throwable) {
+            return null;
+        }
+
+        return $taxon instanceof TaxonInterface ? $taxon : null;
```
Comment (RandyCupic) [inline lib/Item/ValueLoader/TaxonValueLoader.php:27]: @igorkrz Is this check needed? Taxon repository will either return TaxonInterface or null so you can just return the result.

File: lib/Item/ValueLoader/TaxonValueLoader.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
+     */
+    public function __construct(private readonly TaxonRepositoryInterface $taxonRepository) {}
+
+    public function load($id): ?TaxonInterface
+    {
+        try {
+            $taxon = $this->taxonRepository->find($id);
+        } catch (Throwable) {
```
Comment (RandyCupic) [inline lib/Item/ValueLoader/TaxonValueLoader.php:23]: @igorkrz Is this catch necessary?

File: bundle/Resources/config/services/items.yaml [hunk 9]
```diff
@@ -17,3 +17,15 @@ services:
             - "@router"
         tags:
             - { name: netgen_layouts.cms_value_url_generator, value_type: sylius_product }
+
+    netgen_layouts.sylius.item.value_converter.taxon:
```
Comment (emodric) [inline bundle/Resources/config/services/items.yaml:21]: Do taxons have URLs in Sylius? If so, then you're missing the URL generator class.

File: bundle/Resources/config/services/query_types.yaml [hunk 10]
```diff
@@ -22,3 +22,11 @@ services:
             - "@request_stack"
         tags:
             - { name: netgen_layouts.query_type_handler, type: sylius_taxon_products }
+
+    netgen_layouts.sylius.query_type_handler.taxon:
```
Comment (emodric) [inline bundle/Resources/config/services/query_types.yaml]: This should be named with plural: `netgen_layouts.sylius.query_type_handler.taxons`, class too (`TaxonsHandler`), since all other queries use plural

File: lib/Item/ValueConverter/TaxonValueConverter.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+declare(strict_types=1);
+
+namespace Netgen\Layouts\Sylius\Item\ValueConverter;
+
+use Netgen\Layouts\Item\ValueConverterInterface;
+use Sylius\Component\Taxonomy\Model\TaxonInterface;
+
+/**
+ * @implements ValueConverterInterface<TaxonInterface>
```
Comment (emodric) [inline lib/Item/ValueConverter/TaxonValueConverter.php]: PHPDoc need to use FQCN.

File: lib/Item/ValueLoader/TaxonValueLoader.php (hunk 8 above)
Comment (emodric) [reply lib/Item/ValueLoader/TaxonValueLoader.php:23]: Yes it is, it is to ensure that interface does not crash on errors when loading. Instead, the interface will display that the item is invalid and logs need to be checked as to why.

File: lib/Item/ValueLoader/TaxonValueLoader.php [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+use Netgen\Layouts\Item\ValueLoaderInterface;
+use Sylius\Component\Taxonomy\Model\TaxonInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Throwable;
+
+final class TaxonValueLoader implements ValueLoaderInterface
+{
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
```
Comment (emodric) [inline lib/Item/ValueLoader/TaxonValueLoader.php]: FQCN in PHPDoc.

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+use function count;
+use function trim;
+
+class TaxonHandler implements QueryTypeHandlerInterface
+{
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
+     */
+    public function __construct(
+        private readonly TaxonRepositoryInterface $taxonRepository,
```
Comment (emodric) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: Do not use `private readonly`, we're not using this construct here. `private TaxonRepositoryInterface $taxonRepository` is enough.

File: lib/Item/ValueLoader/TaxonValueLoader.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    public function __construct(private readonly TaxonRepositoryInterface $taxonRepository) {}
```
Comment (emodric) [inline lib/Item/ValueLoader/TaxonValueLoader.php]: No `private readonly`.

File: bundle/Resources/config/query_types.yaml (hunk 1 above)
Comment (emodric) [reply bundle/Resources/config/query_types.yaml]: Yes, it should, I also left a similar comment in `services\query_types.yaml`

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+use Netgen\Layouts\Sylius\Parameters\ParameterType as SyliusParameterType;
+use Sylius\Component\Taxonomy\Model\TaxonInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+use function count;
+use function trim;
+
+class TaxonHandler implements QueryTypeHandlerInterface
```
Comment (emodric) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: Needs to be final.

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+{
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
```
Comment (emodric) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: FQCN in PHPDoc.

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 2 above)
Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: @RandyCupic Do you have any other suggestion? `use_current_taxon` is already used in both other queries.

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        );
+
+        $builder->get('use_current_taxon')->add(
+            'taxon_id',
+            SyliusParameterType\TaxonType::class,
+        );
+    }
+
+    /**
+     * @return iterable<array-key, TaxonInterface>
```
Comment (emodric) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: No need for this PHPDoc and it colides with the PHPDoc from the interface.

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 5 above)
Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: @RandyCupic To make PHPStan happy, because `getCurrentRequest` can return `null`.

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 3 above)
Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: @RandyCupic Return type is nullable `?string`.

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+        if (!$taxon instanceof TaxonInterface) {
+            return [];
+        }
+
+        $taxonCode = $taxon->getCode();
+        if ($taxonCode === null) {
+            return [];
+        }
+
+        /** @var TaxonInterface[] $children */
```
Comment (emodric) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: FQCN in PHPDoc.

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 6 above)
Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: @RandyCupic Again, to make PHPStan happy :)

File: lib/Collection/QueryType/Handler/TaxonHandler.php [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+        if (!$taxon instanceof TaxonInterface) {
+            return 0;
+        }
+
+        $taxonCode = $taxon->getCode();
+        if ($taxonCode === null) {
+            return 0;
+        }
+
+        return count($this->taxonRepository->findChildren($taxonCode));
```
Comment (emodric) [inline lib/Collection/QueryType/Handler/TaxonHandler.php]: There isn't any `count` method in the repository? This seems awfully inefficient.

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 4 above)
Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: Maybe, but this follows the implementation in e.g. `LatestProductsHandler`, so we can leave it as is.

**Review by emodric — CHANGES_REQUESTED** [review]

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 2 above)
Comment (RandyCupic) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: @emodric @igorkrz Maybe it's OK then. At first it was not completely clear to me what is this used for. Maybe it should be `Use current taxon as parent` or something like that, but let's keep it for now.

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 3 above)
Comment (RandyCupic) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: In that case OK, otherwise PHP Stan might complain :)

File: lib/Item/ValueLoader/TaxonValueLoader.php (hunk 7 above)
Comment (RandyCupic) [reply lib/Item/ValueLoader/TaxonValueLoader.php:27]: @emodric WDYT?

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 3 above)
Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: All hail PHPStan!

File: lib/Item/ValueLoader/TaxonValueLoader.php (hunk 7 above)
Comment (emodric) [reply lib/Item/ValueLoader/TaxonValueLoader.php:27]: Yes, but PHPStan does not know that 😂

File: bundle/Resources/config/services/items.yaml (hunk 9 above)
Comment (RandyCupic) [reply bundle/Resources/config/services/items.yaml:21]: @emodric @igorkrz They do: https://demo.sylius.com/en_US/taxons/fashion-category/dresses

That's basically the purpose if this PR, so that we can eg. list all taxons as tiles with link to their page (eg. categories).

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 19 above)
Comment (RandyCupic) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: @emodric @igorkrz As far as I know, there is only `count()` in general, but it's not in the `TaxonRepositoryInterface` but deeper. But for `findChildren()` I don't think there's one, so we would need to add it manually.

Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: I already extended `ProductRepository` before to add some needed methods, so we can use the similar approach here.

File: bundle/Resources/config/query_types.yaml (hunk 1 above)
Comment (igorkrz) [reply bundle/Resources/config/query_types.yaml]: Fixed in https://github.com/netgen-layouts/layouts-sylius/pull/19/commits/98dafbcd4286f9b181013dccc0879bac31ec1ac4

File: bundle/Resources/config/services/items.yaml (hunk 9 above)
Comment (igorkrz) [reply bundle/Resources/config/services/items.yaml:21]: Added in https://github.com/netgen-layouts/layouts-sylius/pull/19/commits/98dafbcd4286f9b181013dccc0879bac31ec1ac4

File: bundle/Resources/config/services/query_types.yaml (hunk 10 above)
Comment (igorkrz) [reply bundle/Resources/config/services/query_types.yaml]: Fixed with https://github.com/netgen-layouts/layouts-sylius/pull/19/commits/98dafbcd4286f9b181013dccc0879bac31ec1ac4

File: lib/Item/ValueConverter/TaxonValueConverter.php (hunk 11 above)
Comment (igorkrz) [reply lib/Item/ValueConverter/TaxonValueConverter.php]: Fixed with https://github.com/netgen-layouts/layouts-sylius/pull/19/commits/98dafbcd4286f9b181013dccc0879bac31ec1ac4

File: lib/Item/ValueLoader/TaxonValueLoader.php (hunk 12 above)
Comment (igorkrz) [reply lib/Item/ValueLoader/TaxonValueLoader.php]: Fixed with https://github.com/netgen-layouts/layouts-sylius/pull/19/commits/98dafbcd4286f9b181013dccc0879bac31ec1ac4

File: lib/Collection/QueryType/Handler/TaxonsHandler.php [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        if (!$taxon instanceof TaxonInterface) {
+            return [];
+        }
+
+        $taxonCode = $taxon->getCode();
+        if ($taxonCode === null) {
+            return [];
+        }
+
+        /** @var \Sylius\Component\Taxonomy\Model\TaxonInterface[] $children */
```
Comment (emodric) [inline lib/Collection/QueryType/Handler/TaxonsHandler.php]: This PHPDoc is not needed, `getValues` method is not used anywhere where you explicitly need `TaxonInterface[]` to be typehinted.

File: lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+
+use Netgen\Layouts\Item\ExtendedValueUrlGeneratorInterface;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+/**
+ * @implements \Netgen\Layouts\Item\ExtendedValueUrlGeneratorInterface<\Sylius\Component\Taxonomy\Model\TaxonInterface>
+ */
+final class TaxonValueUrlGenerator implements ExtendedValueUrlGeneratorInterface
+{
+    public function __construct(private readonly UrlGeneratorInterface $urlGenerator) {}
```
Comment (emodric) [inline lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php]: `private readonly` sneaked back in :)

File: lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+            [
+                'slug' => $object->getSlug(),
+            ],
+        );
+    }
+
+    public function generateAdminUrl(object $object): ?string
+    {
+        return $this->urlGenerator->generate(
+            'sylius_admin_taxon_update',
```
Comment (emodric) [inline lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php:30]: Why do you use `update` route? The point is to display the full view of the taxon, both on frontend and on backend.

File: lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+ * @implements \Netgen\Layouts\Item\ExtendedValueUrlGeneratorInterface<\Sylius\Component\Taxonomy\Model\TaxonInterface>
+ */
+final class TaxonValueUrlGenerator implements ExtendedValueUrlGeneratorInterface
+{
+    public function __construct(private readonly UrlGeneratorInterface $urlGenerator) {}
+
+    public function generateDefaultUrl(object $object): ?string
+    {
+        return $this->urlGenerator->generate(
+            'sylius_shop_partial_taxon_show_by_slug',
```
Comment (emodric) [inline lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php:20]: Is this correct? This `partial` part of the route name does not instill confidence :)

File: lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php (hunk 22 above)
Comment (igorkrz) [reply lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php:30]: there is no index or show route in Sylius for taxons(taxa) by default

File: lib/Collection/QueryType/Handler/TaxonsHandler.php (hunk 20 above)
Comment (igorkrz) [reply lib/Collection/QueryType/Handler/TaxonsHandler.php]: This one is tricky, with current code the cs fixer is complaining that it could be a one liner:
return $this->taxonRepository->findChildren($taxonCode);

however - findChildren method in the vendor returns array|TaxonInterface[] without specifying what is returned in array after which the psalm and phpstan are complaining

File: lib/Collection/QueryType/Handler/TaxonHandler.php (hunk 19 above)
Comment (igorkrz) [reply lib/Collection/QueryType/Handler/TaxonHandler.php]: Added in 9ffb3604cfacff8ce7273ad5df453229be482a0f

File: lib/Collection/QueryType/Handler/TaxonsHandler.php (hunk 20 above)
Comment (emodric) [reply lib/Collection/QueryType/Handler/TaxonsHandler.php]: Okay, we will leave it as is.

File: lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php (hunk 21 above)
Comment (igorkrz) [reply lib/Item/ValueUrlGenerator/TaxonValueUrlGenerator.php]: Fixed with b2ba9b2c244e83fb1444dbbee7d70499efebfd08

**Comment by emodric** [comment]
Thanks @igorkrz !
